crates/supervisor/core/src/reorg/handler.rs

```diff
@@ -0,0 +1,156 @@
+use crate::{SupervisorError, rewinder::ChainRewinder};
+use alloy_eips::BlockNumHash;
+use alloy_primitives::{B256, ChainId};
+use alloy_rpc_client::RpcClient;
+use alloy_rpc_types_eth::Block;
+use derive_more::Constructor;
+use kona_protocol::BlockInfo;
+use kona_supervisor_storage::{ChainDb, DerivationStorageReader, HeadRefStorageReader};
+use std::{collections::HashMap, sync::Arc};
+use tracing::{error, info, warn};
+
+/// Handles L1 reorg operations for multiple chains
+#[derive(Debug, Constructor)]
+pub struct ReorgHandler {
+    /// The Alloy RPC client for L1.
+    rpc_client: Arc<RpcClient>,
+    /// Per chain dbs.
+    chain_dbs: HashMap<ChainId, Arc<ChainDb>>,
+}
+
+impl ReorgHandler {
+    /// Processes a reorg for all chains when a new latest L1 block is received
+    pub async fn handle_l1_reorg(&self, latest_block: BlockInfo) -> Result<(), SupervisorError> {
+        info!(
+            target: "reorg_handler",
+            l1_block_number = latest_block.number,
+            "Reorg detected, processing..."
+        );
+
+        let mut failed_chains = Vec::new();
+
+        for (chain_id, chain_db) in self.chain_dbs.iter() {
+            if let Err(err) = self.process_chain_reorg(chain_id, chain_db, latest_block).await {
+                error!(
```

Comment:
is this an error with the supervisor program, or with any of the supervised chains? if the latter, then downgrade to warn log. an error log should indicate that a manual restart of the supervisor is necessary